contract_sale_generation/tests/test_contract_sale.py

```diff
@@ -0,0 +1,114 @@
+        recurring_next_date = to_date("2020-02-15")
+        self.contract.sale_autoconfirm = True
+        self.assertAlmostEqual(self.contract_line.price_subtotal, 50.0)
+        self.contract_line.price_unit = 100.0
+        self.contract.partner_id = self.partner.id
+        self.contract._recurring_create_sale()
+        self.sale_monthly = self.contract._get_related_sales()
+        self.assertTrue(self.sale_monthly)
+        self.assertEqual(self.contract_line.recurring_next_date, recurring_next_date)
+        self.order_line = self.sale_monthly.order_line[0]
+        self.assertTrue(self.order_line.tax_id)
+        self.assertAlmostEqual(self.order_line.price_subtotal, 50.0)
+        self.assertEqual(self.contract.user_id, self.sale_monthly.user_id)
+
+    def test_onchange_contract_template_id(self):
+        """It should change the contract values to match the template."""
+        self.contract.contract_template_id = False
+        self.contract._onchange_contract_template_id()
+        self.contract.contract_template_id = self.template
+        self.contract._onchange_contract_template_id()
+        res = {
+            "contract_type": "sale",
+            "contract_line_ids": [
+                (
+                    0,
+                    0,
+                    {
+                        "product_id": self.product_1.id,
+                        "name": "Test Contract Template",
+                        "quantity": 1,
+                        "uom_id": self.product_1.uom_id.id,
+                        "price_unit": 100,
+                        "discount": 50,
+                        "recurring_rule_type": "yearly",
+                        "recurring_interval": 1,
+                    },
+                )
+            ],
+        }
+        del self.template_vals["name"]
+        self.assertDictEqual(res, self.template_vals)
+
+    def test_contract_count_sale(self):
+        self.contract._recurring_create_sale()
+        self.contract._recurring_create_sale()
+        self.contract._recurring_create_sale()
+        self.contract._compute_sale_count()
+        self.assertEqual(self.contract.sale_count, 3)
+
+    def test_contract_count_sale_2(self):
+        orders = self.env["sale.order"]
+        orders |= self.contract._recurring_create_sale()
+        orders |= self.contract._recurring_create_sale()
+        orders |= self.contract._recurring_create_sale()
+        action = self.contract.action_show_sales()
+        self.assertEqual(set(action["domain"][0][2]), set(orders.ids))
+
+    def test_cron_recurring_create_sale(self):
+        self.contract_line.date_start = "2020-01-01"
+        self.contract_line.recurring_invoicing_type = "post-paid"
+        self.contract_line.date_end = "2020-03-15"
+        self.contract_line._onchange_is_auto_renew()
+        contracts = self.contract2
+        for _i in range(10):
+            contracts |= self.contract.copy({"contract_type": "sale"})
```

Comment:
Currently tests fail at this one line. Error:

`contract/models/contract_line.py", line 452, in _check_recurring_next_date_start_date
    raise ValidationError(
odoo.exceptions.ValidationError: You can't have a date of next invoice anterior to the start of the contract line 'Test Contract Template'
`
Will need to check this more. @NL66278 @thomaspaulb if you have any ideas, please shoot.

Reply:
@ntsirintanis I tried to follow what happens, and somehow it's a problem with the order of the evaluation of computations.

1. The test contract line is set to certain values, among which `date_start = 2020-01-01`
2. The contract including this line is copied 10 times
3. During copying of the contract, `recurring_next_date` of the lines gets recomputed correctly (2020-01-15)
4. Something happens and `recurring_next_date` is now `2020-01-01`, which is lower than 2020-01-15 (maybe recomputation, or goes back to cached value)
5. The computed values are checked and it finds 2020-01-01 < 2020-01-15 and borks out

I find that this stops happening when you access `self.contract_line.recurring_next_date` (by printing, or just like that) before copying, so between line 98 and 99.

Somehow equivalently, @rousseldenis found in [this PR](https://github.com/OCA/contract/pull/841/files#diff-9b7fa04a3606eef1c4322f66d703a6092d2b4bbd783b6b738c9b779dff22731dR106) that he had to call `_compute_recurring_next_date()`, however that specifically does not work for me.

Reply:
There's another failing test on `import odoo.odoo.tests` in another module, I don't know how that got through CI, but it's clearly from and should be `odoo.tests`

Reply:
@thomaspaulb We need first to fix the tests import errors, then, check which test(s) are failing. @ntsirintanis